feat: `@mention` files in codex

moppywhip opened the pull request:
Solves #700

## State of the World Before

Prior to this PR, when users wanted to share file contents with Codex, they had two options:
- Manually copy and paste file contents into the chat
- Wait for the assistant to use the shell tool to view the file

The second approach required the assistant to:
1. Recognize the need to view a file
2. Execute a shell tool call
3. Wait for the tool call to complete
4. Process the file contents

This consumed extra tokens and reduced user control over which files were shared with the model.

## State of the World After

With this PR, users can now:
- Reference files directly in their chat input using the `@path` syntax
- Have file contents automatically expanded into XML blocks before being sent to the LLM

For example, users can type `@src/utils/config.js` in their message, and the file contents will be included in context. Within the terminal chat history, these file blocks will be collapsed back to `@path` format in the UI for clean presentation.

Tag File suggestions:
<img width="857" alt="file-suggestions" src="https://github.com/user-attachments/assets/397669dc-ad83-492d-b5f0-164fab2ff4ba" />

Tagging files in action:
<img width="858" alt="tagging-files" src="https://github.com/user-attachments/assets/0de9d559-7b7f-4916-aeff-87ae9b16550a" />

Demo video of file tagging:
[![Demo video of file tagging](https://img.youtube.com/vi/vL4LqtBnqt8/0.jpg)](https://www.youtube.com/watch?v=vL4LqtBnqt8)

## Implementation Details

This PR consists of 2 main components:

1. **File Tag Utilities**:
   - New `file-tag-utils.ts` utility module that handles both expansion and collapsing of file tags
   - `expandFileTags()` identifies `@path` tokens and replaces them with XML blocks containing file contents
   - `collapseXmlBlocks()` reverses the process, converting XML blocks back to `@path` format for UI display
   - Tokens are only expanded if they point to valid files (directories are ignored)
   - Expansion happens just before sending input to the model

2. **Terminal Chat Integration**:
   - Leveraged the existing file system completion system for tabbing to support the `@path` syntax
   - Added `updateFsSuggestions` helper to manage filesystem suggestions
   - Added `replaceFileSystemSuggestion` to replace input with filesystem suggestions 
   - Applied `collapseXmlBlocks` in the chat response rendering so that tagged files are shown as simple `@path` tags

The PR also includes test coverage for both the UI and the file tag utilities.

## Next Steps

Some ideas I'd like to implement if this feature gets merged:

- Line selection: `@path[50:80]` to grab specific sections of files
- Method selection: `@path#methodName` to grab just one function/class
- Visual improvements: highlight file tags in the UI to make them more noticeable

moppywhip committed:
@ functionality initial work

moppywhip committed:
mvp

moppywhip committed:
added testing

moppywhip committed:
simplified tabbing logic

moppywhip committed:
adding more test cases

moppywhip committed:
simplifying & code cleanup

moppywhip committed:
more cleanup

moppywhip committed:
removing unecessary diffs

moppywhip committed:
removing random space

moppywhip committed:
more cleanup to simplify regex (to not use regex at all)

moppywhip committed:
commonizing logic into a function

moppywhip committed:
suggestions at current workking directory when user types '@' by itself

moppywhip committed:
added terminal chat testing

moppywhip committed:
replacing @-prefix with first file suggestion when pressing enter

moppywhip committed:
moving logic into util file and buffing tests

moppywhip committed:
fixing bug where suggestions would pop out without '@' tag or tabbing

moppywhip committed:
renamed test file and simplified testing mocks

moppywhip committed:
more code cleanup

moppywhip committed:
cleaning up updateFsSuggestions logic

moppywhip committed:
simplifying further

moppywhip committed:
more cleanup

moppywhip committed:
reducing diff size

moppywhip committed:
don't collapse non file xml blocks in chat history

moppywhip committed:
removed redundant test case

LuciNyan commented:
That's awesome! ❤️

moppywhip commented:
@LuciNyan thank you! ^.^

moppywhip commented:
@tibo-openai requesting a review here! I think file tags could be a really nice addition for codex

asadm commented:
this would be amazing! thanks.

Later, I wonder if we can also support entire folders (and pack them as xml along with a dir structure at top). This would be great with models that have 1M+ context window. Right now, I do that kind of operations manually with a custom packer tool I made.

moppywhip committed:
Merge branch 'main' of https://github.com/openai/codex into @-tag
updating branch with commits from main.

moppywhip committed:
clarified comment

fouad-openai reviewed (approved):
thanks for this change @moppywhip - really excited to `@mention` files!

there's a few improvements we likely want to make around the UX like pressing enter while selecting files should continue the path rather than sending (like how `tab` in shell typically works)

moppywhip commented:
@fouad-openai  I'm very happy to hear that the feature looks good! 

> there's a few improvements we likely want to make around the UX like pressing enter while selecting files should continue the path rather than sending (like how tab in shell typically works)

I wanted to double check if my understanding is correct here. 

Say we are trying to tag ``src/app.tsx``, and we've already typed out ``@s``. 

In the terminal, we should see something like: 

<img width="853" alt="Screenshot 2025-04-29 at 2 42 39 PM" src="https://github.com/user-attachments/assets/70c161c6-1bfe-48c6-9167-65e462480345" />

If we press enter here, we should see something like: 

<img width="863" alt="Screenshot 2025-04-29 at 2 43 02 PM" src="https://github.com/user-attachments/assets/f40c0caa-7830-46fb-9cb7-a2c19451c786" />

Am I correct in this interpretation?

fouad-openai commented:
yes! exactly, and if I press enter on this image it should complete the path in the text input before sending

![image](https://github.com/user-attachments/assets/143bcd8b-bbe8-4d78-8280-1ee57cdffe34)

moppywhip committed:
mvp of uiux changes

moppywhip committed:
finalizing tests

moppywhip committed:
cleaning up comments and factoring out common logic

moppywhip committed:
more comment cleanup

moppywhip committed:
typo

moppywhip committed:
updating more comments

fouad-openai commented:
just tested latest and it seems very close! I think last thing is to complete file path + close file mention selector, instead of sending the message - does that seem right @moppywhip?

moppywhip commented:
@fouad-openai 

so sorry abt lack of communication. I ended up finding a pretty nasty race condition bug and have been fixing it. will be putting up the commit in < 30 mins.

fouad-openai commented:
ah no worries! just wanted to make sure you weren't blocked on me. thank you again for the contribution 🙏

moppywhip committed:
fixed breaking bug where multiple hooks were racing to call setFsSuggestions()

moppywhip commented:
@fouad-openai 
Bugs are fixed! 

Now, if the current file system selection is a directory, pressing enter will simply replace the partial tag with the directory. 
If the selection is a file, pressing enter will replace the partial tag and submit. 

here's a quick video demoing the changes:

https://github.com/user-attachments/assets/d33e3858-25bd-4278-89b1-6143efa61475

Also, there's been a slight regression that I've been having some trouble fixing. You can see it in the video but basically when you fill a partial tag with the selection, the cursor will snap to the beginning of the line for a brief second before moving to the end. I'm still trying to figure it out on my end but any input would be super appreciated 🙏

moppywhip commented:
I do apologize but I might need to log off now since I have finals lol. Will try to fix this tmrw!

moppywhip committed:
clarified comment

moppywhip committed:
updating branch with upstream changes

moppywhip committed:
fixing cursor snapback issue

moppywhip committed:
undoing unintended change

moppywhip commented:
hi @fouad-openai, I ended up fixing the issue. the problem was that the underlying multi-line editor component was remounting (and resetting the cursor to the beginning!) unnecessarily when we replaced the input text with a suggestion. 

I fixed this by adding a new method to update the editor's text content and cursor position directly via its ref, so we don't have to remount and move the cursor.

moppywhip commented:
looks much better now! 

https://github.com/user-attachments/assets/72a57afb-17ac-43e7-81c6-72abb17a572e

moppywhip committed:
fixing small bug with fs suggestions and backspace

moppywhip committed:
updating setText impl to be more idiomatic w.r.t everything else

moppywhip committed:
Merge branch 'main' of https://github.com/openai/codex into @-tag

moppywhip committed:
improved naming for helper function

moppywhip committed:
much cleaner implementation of cursor movements

moppywhip committed:
reducing diffs

moppywhip committed:
added tests to ensure correctness in cursor initialization logic

moppywhip committed:
cleaned up text buffer construction logic

moppywhip committed:
cleaned up setCursorIdx

fouad-openai commented:
ok great, thank you!
